Firestore/core/src/firebase/firestore/remote/online_state_tracker.mm

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/remote/online_state_tracker.h"
+
+#include <chrono>  // NOLINT(build/c++11)
+
+#import "Firestore/Source/Remote/FSTRemoteStore.h"
+
+#include "Firestore/core/src/firebase/firestore/util/executor.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "Firestore/core/src/firebase/firestore/util/string_format.h"
+
+namespace chr = std::chrono;
+using firebase::firestore::model::OnlineState;
+using firebase::firestore::util::AsyncQueue;
+using firebase::firestore::util::DelayedOperation;
+using firebase::firestore::util::Status;
+using firebase::firestore::util::StringFormat;
+using firebase::firestore::util::TimerId;
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace {
+
+// To deal with transient failures, we allow multiple stream attempts before
+// giving up and transitioning from OnlineState Unknown to Offline.
+// TODO(mikelehen): This used to be set to 2 as a mitigation for b/66228394.
+// @jdimond thinks that bug is sufficiently fixed so that we can set this back
+// to 1. If that works okay, we could potentially remove this logic entirely.
+const int kMaxWatchStreamFailures = 1;
+
+// To deal with stream attempts that don't succeed or fail in a timely manner,
+// we have a timeout for OnlineState to reach Online or Offline. If the timeout
+// is reached, we transition to Offline rather than waiting indefinitely.
+const AsyncQueue::Milliseconds kOnlineStateTimeout = chr::seconds(10);
+
+}  // namespace
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+void OnlineStateTracker::HandleWatchStreamStart() {
+  if (watch_stream_failures_ != 0) {
```

Comment:
This is inconsistent with other platforms, but I personally feel an early return is very helpful here, given significant indentation levels.

Reply:
I have no problem with doing early return but we should try to keep those other platforms in sync. If you want to defer that other work for now perhaps it's worth using the other form for now or creating a `TODO(var-const)` to port to the other platforms?

Reply:
I'll tweak the other two platforms this week, so no need for a TODO.